### PR TITLE
Add new setting to disable persistent tasks allocations

### PR DESCRIPTION
**Pull request by tlrx**
This commit adds a new setting `cluster.persistent_tasks.allocation.enable` that can be used to enable or disable the allocation of persistent tasks. 

The setting accepts the values `all` (default) or `none`. When set to none, the persistent tasks that are created (or that must be reassigned) won't be assigned to a node but will reside in the cluster state with
a `null` executor node and a reason describing why it is not assigned:

```
"assignment" : {
  "executor_node" : null,
  "explanation" : "persistent task [foo/bar] cannot be assigned [no persistent task assignments are allowed due to cluster settings]"
}
```

This commit adds some AssignmentDecider and AssignmentDecision classes that are similar to the deciders/decisions classes used for shard allocations. It also adds an EnableAssignmentDecider decider that is responsible for registering the new cluster setting.

The setting is checked when executing the `PersistentTasksClusterService#createAssignment()` method, before delegating the creation of the assignment to the PersistentTasksExecutor. As a follow up, I think that PersistentTasksExecutor could expose a `getTaskAllocationDecider` method that would be used by the persistent task cluster service to build a list of nodes that could execute the new task. Then this list of assignable nodes could be passed again to the executor to select the exact node to execute the task.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * {@link AssignmentDecision} represents the decision made during the process of
+ * assigning a persistent task to a node of the cluster.
+ *
+ * @see AssignmentDecider
+ */
+public final class AssignmentDecision {
+
+    public static final AssignmentDecision ALWAYS = new AssignmentDecision(Type.YES);
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java]: That feels like premature optimization.

**Review by imotov — APPROVED** [review]
Could you open an issue for the follow-up that you proposed at the end of your PR so it wouldn't get lost? Without such follow-up this PR doesn't make a lot of sense. Ideally, I would have preferred for this follow-up to happen first and this change to follow, but it might be not possible if we need this for 6.3.

**Comment by tlrx** [comment]
> Could you open an issue for the follow-up that you proposed at the end of your PR so it wouldn't get lost? Without such follow-up this PR doesn't make a lot of sense. Ideally, I would have preferred for this follow-up to happen first and this change to follow, but it might be not possible if we need this for 6.3.

Yes, the follow up is a bigger change and we'd like the setting to be added in 6.3, thus the order of pull requests. Thanks for your review Igor.

File: server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java [hunk 2]
```diff
@@ -45,12 +48,14 @@
 
     private final ClusterService clusterService;
     private final PersistentTasksExecutorRegistry registry;
+    private final AssignmentDecider<PersistentTaskParams> decider;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java]: can we just hard code this for now to EnableAssignmentDecider? we can generalize when it becomes relevant (and know what we want exactly)

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecider.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.persistent.PersistentTaskParams;
+
+/**
+ * {@link AssignmentDecider} are used to decide if a persistent
+ * task can be assigned to a node.
+ */
+public abstract class AssignmentDecider<Params extends PersistentTaskParams> extends AbstractComponent {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecider.java]: I would prefer this to be added at the same time we support multiple deciders.

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java]: +1 . People can just use yes? I'm not sure what the difference is?

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public static final AssignmentDecision YES = new AssignmentDecision(Type.YES);
+    public static final AssignmentDecision NO = new AssignmentDecision(Type.NO);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java]: I'm not sure we want this? it means people can say no without giving a reason.

File: server/src/main/java/org/elasticsearch/persistent/decider/EnableAssignmentDecider.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * <p>
+ * Allocation settings can have the following values (non-casesensitive):
+ * <ul>
+ *     <li> <code>NONE</code> - no persistent tasks can be assigned
+ *     <li> <code>ALL</code> - all persistent tasks can be assigned to nodes
+ * </ul>
+ *
+ * @see Allocation
+ */
+public class EnableAssignmentDecider<Params extends PersistentTaskParams> extends AssignmentDecider<Params> {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/persistent/decider/EnableAssignmentDecider.java]: Why is this generic?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @tlrx . I left some suggestions, mostly about removing stuff that are not yet needed.

I would love to have an integration test (i.e., qa) to check that what we need (setting this in the es.yml file actually prevents initial allocation of tasks and that you can enable them from the API). This may be tricky so we need to check whether we can do this in time.

Re your suggestion of opening up the deciders to the executors - I'll reach you to talk in another channel. I want to understand better where you see this heading

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java]: Right, it's not required.

File: server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java (hunk 4 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/persistent/decider/AssignmentDecision.java]: Right

**Comment by tlrx** [comment]
Thanks @imotov and @bleskes and sorry for all the premature stuff.

@bleskes I addressed your comments and also added an integration test (but not a QA one), please let me know what you think.

**Review by bleskes — APPROVED** [review]
LGTM. Thanks @tlrx

**Comment by tlrx** [comment]
Thanks @bleskes

**Pull request merged**